
# 📝  Evidencia de Aprendizaje 1 - Creación de una base de datos analítica


**Autora:** Aida Luz Aguirre hernandez 

**Curso:** Big Data

**Grupo:** grupo_8 

---

# Proyecto: Creación de una Base de Datos Analítica para el Análisis de Ventas en un Negocio de Comercio Electrónico

## 1️⃣ Problema y Dataset

### Empresa y Contexto

**Empresa:** TechStore Online S.A. (ficticia)  
**Sector:** Comercio electrónico / Retail tecnológico  
**Equipo destinatario:** Gerencia de ventas y equipo de análisis de datos.  
**Rol de mi actividad:** Crear una Base de datos Analítica.  

**Contexto del problema:**  
En el contexto actual del comercio electrónico, las empresas manejan grandes volúmenes de información provenientes de sus operaciones diarias: pedidos, productos, clientes y países de destino. Sin una base de datos estructurada y analítica, resulta complejo identificar patrones de compra, evaluar el desempeño por país o producto y optimizar la toma de decisiones comerciales.
TechStore Online vende productos tecnológicos mediante su plataforma en línea. Actualmente, los datos de ventas, clientes y productos se encuentran dispersos en archivos planos y hojas de cálculo, lo que dificulta su consulta, consolidación y análisis. Antes de que el equipo de análisis de negocio pueda generar insights o reportes de ventas, marketing e inventario, es necesario **estructurar, limpiar y cargar la información en una base de datos analítica** confiable y consistente.

**Objetivo de la migración de datos:**  
1. Consolidar los registros históricos de ventas, clientes y productos en un sistema centralizado.  
2. Garantizar integridad y consistencia de los datos, respetando claves primarias y relaciones entre tablas.  
3. Preparar los datos para que el equipo de análisis pueda realizar consultas, segmentación y generación de KPIs sin necesidad de manipular los archivos originales. 

---
### Objetivos del proyecto

**Objetivo general:**  
Migrar y estructurar los datos del negocio online de TechStore en una base de datos analítica que permita su posterior análisis por el equipo de negocio.  

**Objetivos específicos:**  
- Diseñar una estructura de base de datos relacional que relacione clientes, productos y ventas.  
- Cargar los datos del dataset de manera correcta, garantizando integridad y consistencia.  
- Crear la base de datos relacional en SQLite y cargar los datos del dataset “An Online Shop Business”.  
- Ejecutar consultas SQL que evidencien la correcta creación y comprensión de la base de datos.
- Interpretar los resultados de las consultas para obtener hallazgos analíticos relevantes sobre las ventas.

---

### Dataset

**Nombre del dataset:** An Online Shop Business  
**Autor en Kaggle:** Gabriel Ramos  
**Enlace:** [E-commerce Business Transaction](https://www.kaggle.com/datasets/gabrielramos87/an-online-shop-business) 


**Descripción:**  
- Columnas principales: número de transacción, código de producto, descripción, cantidad, fecha de venta, precio unitario, identificador de cliente y país del cliente.  
- Ideal para la construcción de una base de datos relacional que soporte análisis de ventas, clientes y productos.  

**Justificación de selección:**  
Elegí este dataset porque refleja un escenario realista de comercio electrónico y me permite practicar todo el proceso de migración y preparación de datos que es el objetivo de mi proyecto. Contiene información completa sobre ventas, productos y clientes, lo que permite crear una estructura de base de datos relacional que relacione correctamente estas tablas. Además, tiene un tamaño considerable, lo que me permite trabajar con un volumen de datos realista, similar al que podría encontrar en una empresa de comercio electrónico, y realizar pruebas de integridad y consistencia, ya que incluye claves primarias, claves foráneas y datos detallados de cada transacción. También me da la oportunidad de preparar y limpiar los datos antes de cargarlos a una base de datos analítica. Finalmente, me resulta motivador porque puedo ver cómo un conjunto de datos desorganizado y disperso se transforma en información estructurada y lista para análisis, reflejando un problema real que enfrentan muchas empresas de comercio electrónico. Este dataset me permite aplicar de manera práctica los conceptos de migración, limpieza y preparación de datos, trabajando con información que simula un escenario real y que tiene suficiente complejidad para aprender y demostrar mis habilidades.

---

### 📋 Variables relevantes y su utilidad

| Variable (Inglés) | Variable (Español) | Tipo de dato | Descripción / Utilidad en la base analítica |
|------------------|-------------------|--------------|--------------------------------------------|
| transaction_id   | ID de transacción | Texto | Identificador único de cada venta; clave primaria en la tabla **Sales**. |
| product_id       | ID de producto | Texto | Código único del producto; clave primaria en **Products** y clave foránea en **Sales**. |
| product_name     | Nombre del producto | Texto | Nombre o descripción del producto; permite identificar los artículos más vendidos. |
| unit_price       | Precio unitario | Decimal | Valor unitario del producto; usado para cálculos de ingresos y análisis financiero. |
| quantity         | Cantidad | Entero | Número de unidades vendidas; permite calcular totales y analizar la demanda. |
| date             | Fecha de venta | Texto / Fecha | Fecha en que ocurrió la transacción; útil para análisis temporal o estacional. |
| customer_id      | ID de cliente | Entero | Identificador único de cada cliente; clave primaria en **Customers** y foránea en **Sales**. |
| country          | País | Texto | País del cliente; permite segmentar las ventas por ubicación geográfica. |

---

### Interpretación

Estas variables se organizan en tres tablas principales que conforman una base de datos relacional analítica:

| Tabla (Español) | Tabla (Inglés) | Campos (Español) | Campos (Inglés) | Descripción |
|-----------------|----------------|------------------|----------------|-------------|
| **Clientes** | `Customers` | `IDCliente`, `País` | `customer_id`, `country` | Contiene la información básica de los clientes, incluyendo su identificador único y país. |
| **Productos** | `Products` | `IDProducto`, `NombreProducto`, `PrecioUnitario` | `product_id`, `product_name`, `unit_price` | Incluye el catálogo de productos y su precio unitario. |
| **Ventas** | `Sales` | `IDTransacción`, `IDCliente`, `IDProducto`, `Cantidad`, `Fecha` | `transaction_id`, `customer_id`, `product_id`, `quantity`, `date` | Registra todas las transacciones realizadas, conectando clientes con productos. |


Este modelo facilita consultas analíticas para:
- Determinar los productos más vendidos.  
- Analizar ventas por país o cliente.  
- Identificar patrones de compra y estacionalidad.  
- Calcular métricas clave como ingresos totales, ticket promedio o volumen por producto.

---

 

El objetivo final de esta etapa es garantizar que los datos estén limpios, correctamente tipados y listos para ser consumidos por el equipo de análisis sin necesidad de manipulación adicional de archivos originales.


## 2️⃣ Modelo Entidad–Relación (ERD)

### Diagrama

Puedes ver el diagrama completo en dbdiagram.io en este enlace:  
[ERD TechStore Online S.A.](https://dbdiagram.io/d/modelo-entidad-relacion-ERD_TechStore-Online-S-A-690d69576735e11170aa4e58)

![Diagrama ERD](ERD_TechStore Online S.A.png)








    CUSTOMER {
        float customer_id PK
        string country
    }

    PRODUCT {
        string product_id PK
        string product_name
        float unit_price
    }

    SALES {
        string transaction_id PK
        string date
        float customer_id FK
        string product_id FK
        int quantity
    }

    CUSTOMER ||--o{ SALES : "realiza"
    PRODUCT ||--o{ SALES : "contiene"



### Explicación del modelo ERD

**1.Tabla Customer**

- customer_id: Clave primaria que identifica de manera única a cada cliente.

- country: País del cliente, útil para análisis geográfico y segmentación de mercado.

Descripción:
Esta tabla almacena la información de los clientes que realizaron compras. Sirve como base para relacionar las ventas con su ubicación geográfica y analizar comportamientos de compra por región.

**2.Tabla Product**

- product_id: Clave primaria que identifica de manera única cada producto.

- product_name: Nombre o descripción del producto.

- unit_price: Precio unitario del producto, necesario para calcular ingresos y márgenes.

Descripción:
Contiene la información de los productos disponibles en la tienda online.
Permite relacionar las transacciones con los artículos vendidos y analizar la demanda por producto o categoría.

**3.Tabla Sales**

- transaction_id: Clave primaria que identifica cada transacción de venta.

- customer_id: Clave foránea que se relaciona con Customer.customer_id.

- product_id: Clave foránea que se relaciona con Product.product_id.

- quantity: Cantidad de unidades vendidas en la transacción.

- date: Fecha en que se realizó la venta.

Descripción:
Registra cada operación comercial, asociando un cliente con un producto específico en una fecha determinada.
Esta tabla es el núcleo del modelo, ya que permite realizar análisis de ventas, ticket promedio, comportamiento de clientes y rendimiento de productos.

### Relaciones

- **Customer → Sales:** Un cliente puede realizar muchas ventas (`1:N`), pero cada venta corresponde a un solo cliente.  
- **Product → Sales:** Un producto puede aparecer en muchas ventas (`1:N`), y cada venta contiene un producto específico.  

Este modelo relacional refleja la estructura del dataset “An Online Shop Business”.
Asegura que los datos estén organizados de manera lógica y consistente, estableciendo la base para la creación de la base de datos real y la ejecución de consultas analíticas a futuro.

## 3️⃣ Creación de la base de datos y carga de información

Se creó la base de datos SQLite tienda_online.db con tres tablas: Customer (customer_id, country), Product (product_id, product_name, unit_price) y Sales (transaction_id, date, product_id, customer_id, quantity). Los datos del dataset se cargaron en pandas, se limpiaron valores nulos, se renombraron las columnas para coincidir con el modelo relacional y se insertaron en SQLite usando to_sql(). Finalmente, se verificó la correcta carga de la información consultando los primeros registros de cada tabla y realizando consultas de ejemplo para confirmar las relaciones entre clientes, productos y ventas.

### Instalación de librerías y descarga del dataset

In [0]:
# Librerías necesarias
import os
import json
import zipfile
from pathlib import Path
from typing import List, Any

import numpy as np
import pandas as pd
import kagglehub

# Descargar dataset desde Kaggle
dataset_path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
print("Ruta a los archivos del dataset:", dataset_path)


Ruta a los archivos del dataset: /home/spark-eca814ff-d0c9-45f8-afff-a7/.cache/kagglehub/datasets/gabrielramos87/an-online-shop-business/versions/7


### Cargar el dataset como DataFrame de pandas

In [0]:
# Revisar si hay archivos zip, csv o xlsx en la ruta descargada
archivos = os.listdir(dataset_path)
print("Archivos descargados:", archivos)

# Si hay zip, extraer
zip_files = [f for f in archivos if f.endswith(".zip")]
if zip_files:
    zip_file = os.path.join(dataset_path, zip_files[0])
    extract_dir = os.path.join(dataset_path, "extracted")
    os.makedirs(extract_dir, exist_ok=True)
    with zipfile.ZipFile(zip_file, "r") as z:
        z.extractall(extract_dir)
    data_dir = extract_dir
else:
    # Si no hay zip, usar la ruta descargada directamente
    data_dir = dataset_path

print("Ruta final de datos:", data_dir)


Archivos descargados: ['Sales Transaction v.4a.csv']
Ruta final de datos: /home/spark-eca814ff-d0c9-45f8-afff-a7/.cache/kagglehub/datasets/gabrielramos87/an-online-shop-business/versions/7


In [0]:
# Buscar archivos CSV o XLSX
archivos = os.listdir(data_dir)
csv_files = [f for f in archivos if f.endswith(".csv")]
xlsx_files = [f for f in archivos if f.endswith(".xlsx")]

# Cargar CSV
if csv_files:
    df = pd.read_csv(os.path.join(data_dir, csv_files[0]), encoding="latin1")
elif xlsx_files:
    df = pd.read_excel(os.path.join(data_dir, xlsx_files[0]))
else:
    raise FileNotFoundError("No se encontró archivo CSV ni XLSX en el dataset")

# Mostrar las primeras filas
df.head()


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,12/9/2019,22485,Set Of 2 Wooden Market Crates,21.47,12,17490.0,United Kingdom
1,581475,12/9/2019,22596,Christmas Star Wish List Chalkboard,10.65,36,13069.0,United Kingdom
2,581475,12/9/2019,23235,Storage Tin Vintage Leaf,11.53,12,13069.0,United Kingdom
3,581475,12/9/2019,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069.0,United Kingdom
4,581475,12/9/2019,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069.0,United Kingdom


### Limpieza básica de datos

In [0]:
# Revisar valores nulos
df.isnull().sum()

# Reemplazar nulos
df = df.fillna({
    "CustomerNo": 0,
    "ProductNo": "ND",
    "ProductName": "ND",
    "Price": 0.0,
    "Country": "ND",
    "Date": "1900-01-01" # fecha por defecto si falta
})

# Renombrar columnas para profesionalizar el dataset
df.rename(columns={
    'TransactionNo': 'transaction_id',
    'Date': 'date',
    'ProductNo': 'product_id',
    'ProductName': 'product_name',
    'Price': 'unit_price',
    'Quantity': 'quantity',
    'CustomerNo': 'customer_id',
    'Country': 'country'
}, inplace=True)

print("✅ Columnas renombradas correctamente:")
print(df.columns.tolist())





✅ Columnas renombradas correctamente:
['transaction_id', 'date', 'product_id', 'product_name', 'unit_price', 'quantity', 'customer_id', 'country']


### Creación de la Base de Datos e Inserción de Información en SQLite

In [0]:
import sqlite3
import pandas as pd

# Crear o conectar a la base de datos local
conn = sqlite3.connect("tienda_online.db")

# Crear tablas basadas en el modelo ER profesional
df_customers = df[['customer_id', 'country']].drop_duplicates().copy()
df_products = df[['product_id', 'product_name', 'unit_price']].drop_duplicates().copy()
df_sales = df[['transaction_id', 'date', 'product_id', 'customer_id', 'quantity']].copy()

# Guardar las tablas en SQLite
df_customers.to_sql("Customer", conn, if_exists="replace", index=False)
df_products.to_sql("Product", conn, if_exists="replace", index=False)
df_sales.to_sql("Sales", conn, if_exists="replace", index=False)

print("✅ Base de datos 'tienda_online.db' creada exitosamente con nombres profesionalizados.")


✅ Base de datos 'tienda_online.db' creada exitosamente con nombres profesionalizados.


In [0]:
print("📋 Tablas en la base de datos:")
print(pd.read_sql("SELECT name FROM sqlite_master WHERE type='table';", conn))


📋 Tablas en la base de datos:
       name
0  Customer
1   Product
2     Sales


In [0]:
print("📋 Vista previa de cada tabla:")

print("\n🧍 Customer:")
display(pd.read_sql("SELECT * FROM Customer LIMIT 5;", conn))

print("\n📦 Product:")
display(pd.read_sql("SELECT * FROM Product LIMIT 5;", conn))

print("\n💳 Sales:")
display(pd.read_sql("SELECT * FROM Sales LIMIT 5;", conn))


📋 Vista previa de cada tabla:

🧍 Customer:


customer_id,country
17490.0,United Kingdom
13069.0,United Kingdom
12433.0,Norway
13426.0,United Kingdom
17364.0,United Kingdom



📦 Product:


product_id,product_name,unit_price
22485,Set Of 2 Wooden Market Crates,21.47
22596,Christmas Star Wish List Chalkboard,10.65
23235,Storage Tin Vintage Leaf,11.53
23272,Tree T-Light Holder Willie Winkie,10.65
23239,Set Of 4 Knick Knack Tins Poppies,11.94



💳 Sales:


transaction_id,date,product_id,customer_id,quantity
581482,12/9/2019,22485,17490.0,12
581475,12/9/2019,22596,13069.0,36
581475,12/9/2019,23235,13069.0,12
581475,12/9/2019,23272,13069.0,12
581475,12/9/2019,23239,13069.0,6


### 4️⃣ Evidencia con consultas SQL

A continuación se realiozan algunas consultas SQL que permiten evidenciar la correcta creación y carga de la base de datos tienda_online.db y sus tablas (Customer, Product y Sales).

**1. Conteo de registros por tabla**: Para verificar cuántos registros contiene cada tabla.

In [0]:
import pandas as pd

print("🧍 Tabla Customer:")
display(pd.read_sql("SELECT COUNT(*) AS total_registros FROM Customer;", conn))

🧍 Tabla Customer:


total_registros
4739


Interpretación: La tabla Customer contiene 4,739 clientes únicos. Esto confirma que se cargaron correctamente los datos de los clientes sin duplicados.

In [0]:
print("📦 Tabla Product:")
display(pd.read_sql("SELECT COUNT(*) AS total_registros FROM Product;", conn))



📦 Tabla Product:


total_registros
23631


Interpretación: La tabla Product tiene 23,631 productos únicos. Esto evidencia que todos los productos del dataset se migraron correctamente a la base de datos.

In [0]:
print("💳 Tabla Sales:")
display(pd.read_sql("SELECT COUNT(*) AS total_registros FROM Sales;", conn))

💳 Tabla Sales:


total_registros
536350


Interpretación: La tabla Sales contiene 536,350 transacciones, lo que refleja la totalidad de las ventas registradas en el dataset. Esto confirma que todas las operaciones se cargaron correctamente y se relacionan con los clientes y productos correspondientes.

**2. Nombres y tipos de columnas:** Para confirmar que las tablas fueron creadas según el modelo relacional definido

SQLite: PRAGMA table_info(nombre_tabla);

In [0]:
print("🧍 Tabla Customer:")
display(pd.read_sql("PRAGMA table_info('Customer');", conn))

🧍 Tabla Customer:


cid,name,type,notnull,dflt_value,pk
0,customer_id,REAL,0,null,0
1,country,TEXT,0,null,0


Columnas:

customer_id (INTEGER) → Clave primaria que identifica de forma única a cada cliente.

country (TEXT) → País del cliente, útil para segmentación geográfica.

Interpretación: La tabla Customer tiene las columnas esperadas según el modelo ERD y los tipos de datos son consistentes con los valores que contiene.


In [0]:
print("📦 Tabla Product:")
display(pd.read_sql("PRAGMA table_info('Product');", conn))

📦 Tabla Product:


cid,name,type,notnull,dflt_value,pk
0,product_id,TEXT,0,null,0
1,product_name,TEXT,0,null,0
2,unit_price,REAL,0,null,0


Columnas:

product_id (TEXT) → Clave primaria de cada producto.

product_name (TEXT) → Nombre o descripción del producto.

unit_price (REAL) → Precio unitario del producto.

Interpretación: La tabla Product refleja correctamente la estructura esperada. Los tipos de datos permiten realizar cálculos sobre precios y relacionar productos con ventas.

In [0]:
print("💳 Tabla Sales:")
display(pd.read_sql("PRAGMA table_info('Sales');", conn))


💳 Tabla Sales:


cid,name,type,notnull,dflt_value,pk
0,transaction_id,TEXT,0,null,0
1,date,TEXT,0,null,0
2,product_id,TEXT,0,null,0
3,customer_id,REAL,0,null,0
4,quantity,INTEGER,0,null,0


Columnas:

transaction_id (TEXT) → Clave primaria de cada transacción.

date (TEXT/TIMESTAMP) → Fecha de la venta.

product_id (TEXT) → Relación con la tabla Product.

customer_id (INTEGER) → Relación con la tabla Customer.

quantity (INTEGER) → Cantidad de unidades vendidas.

Interpretación: La tabla Sales contiene todas las columnas necesarias para registrar cada transacción y relacionarla con clientes y productos. Los tipos de datos son adecuados para realizar análisis de ventas, agregaciones y filtros por fecha, cliente o producto.

**4. Consultas con filtros:** Para evidenciar la correcta relación entre tablas y la posibilidad de extraer información específica:

In [0]:
### Consulta filtrada: Ventas de clientes en Noruega
query = """
SELECT DISTINCT s.transaction_id, s.date, s.quantity, p.product_name, c.country
FROM Sales s
JOIN Product p ON s.product_id = p.product_id
JOIN Customer c ON s.customer_id = c.customer_id
WHERE c.country = 'Norway'
LIMIT 10;
"""
display(pd.read_sql(query, conn))


transaction_id,date,quantity,product_name,country
536532,12/1/2018,2,Doormat Red Retrospot,Norway
536532,12/1/2018,36,Set/20 Red Retrospot Paper Napkins,Norway
536532,12/1/2018,8,Retrospot Childrens Apron,Norway
536532,12/1/2018,12,Cook With Wine Metal Sign,Norway
536532,12/1/2018,24,Pack Of 72 Retrospot Cake Cases,Norway
536532,12/1/2018,24,Pack Of 72 Skull Cake Cases,Norway
536532,12/1/2018,12,First Aid Tin,Norway
536532,12/1/2018,12,Fancy Font Birthday Card,Norway
536532,12/1/2018,12,Booze & Women Greeting Card,Norway
536532,12/1/2018,6,Red Hanging Heart T-Light Holder,Norway


Interpretación:

- Esta consulta muestra las ventas realizadas por clientes de Noruega, combinando información de las tablas Sales, Product y Customer.

- Se observa que una misma transacción (transaction_id = 536532) puede contener múltiples productos, lo que refleja la relación 1:N entre ventas y productos.

- Los datos incluyen la fecha de la venta, cantidad vendida, nombre del producto y país del cliente, evidenciando que las tablas están correctamente relacionadas.

- Esto confirma que la base de datos fue cargada correctamente y que es posible realizar filtros y análisis por país, cliente o producto.